src/main/drivers/dma_stm32h7xx.c

```diff
@@ -0,0 +1,148 @@
+    DEFINE_DMA_CHANNEL(DMA1, 5, 38),
+    DEFINE_DMA_CHANNEL(DMA1, 6, 48),
+    DEFINE_DMA_CHANNEL(DMA1, 7, 54),
+
+    DEFINE_DMA_CHANNEL(DMA2, 0,  0),
+    DEFINE_DMA_CHANNEL(DMA2, 1,  6),
+    DEFINE_DMA_CHANNEL(DMA2, 2, 16),
+    DEFINE_DMA_CHANNEL(DMA2, 3, 22),
+    DEFINE_DMA_CHANNEL(DMA2, 4, 32),
+    DEFINE_DMA_CHANNEL(DMA2, 5, 38),
+    DEFINE_DMA_CHANNEL(DMA2, 6, 48),
+    DEFINE_DMA_CHANNEL(DMA2, 7, 54),
+
+};
+
+/*
+ * DMA IRQ Handlers
+ */
+DEFINE_DMA_IRQ_HANDLER(1, 0, DMA1_ST0_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 1, DMA1_ST1_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 2, DMA1_ST2_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 3, DMA1_ST3_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 4, DMA1_ST4_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 5, DMA1_ST5_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 6, DMA1_ST6_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(1, 7, DMA1_ST7_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 0, DMA2_ST0_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 1, DMA2_ST1_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 2, DMA2_ST2_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 3, DMA2_ST3_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 4, DMA2_ST4_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 5, DMA2_ST5_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 6, DMA2_ST6_HANDLER)
+DEFINE_DMA_IRQ_HANDLER(2, 7, DMA2_ST7_HANDLER)
+
+static void enableDmaClock(int index)
+{
+    // This is essentially copies of __HAL_RCC_DMA{1,2}_CLK_ENABLE macros
+    // squashed into one.
+
+    const uint32_t rcc = dmaDescriptors[index].dma == DMA1 ? RCC_AHB1ENR_DMA1EN : RCC_AHB1ENR_DMA2EN;
+
+    do {
+        __IO uint32_t tmpreg;
+        SET_BIT(RCC->AHB1ENR, rcc);
+        /* Delay after an RCC peripheral clock enabling */
+        tmpreg = READ_BIT(RCC->AHB1ENR, rcc);
+        UNUSED(tmpreg);
+    } while (0);
+}
+
+void dmaInit(dmaIdentifier_e identifier, resourceOwner_e owner, uint8_t resourceIndex)
+{
+    const int index = DMA_IDENTIFIER_TO_INDEX(identifier);
+
+    enableDmaClock(index);
+    dmaDescriptors[index].owner = owner;
+    dmaDescriptors[index].resourceIndex = resourceIndex;
+}
+
+void dmaSetHandler(dmaIdentifier_e identifier, dmaCallbackHandlerFuncPtr callback, uint32_t priority, uint32_t userParam)
+{
+    const int index = DMA_IDENTIFIER_TO_INDEX(identifier);
+
+    enableDmaClock(index);
+    dmaDescriptors[index].irqHandlerCallback = callback;
+    dmaDescriptors[index].userParam = userParam;
+
+    HAL_NVIC_SetPriority(dmaDescriptors[index].irqN, NVIC_PRIORITY_BASE(priority), NVIC_PRIORITY_SUB(priority));
+    HAL_NVIC_EnableIRQ(dmaDescriptors[index].irqN);
+}
+
+resourceOwner_e dmaGetOwner(dmaIdentifier_e identifier)
+{
+    return dmaDescriptors[DMA_IDENTIFIER_TO_INDEX(identifier)].owner;
+}
+
+uint8_t dmaGetResourceIndex(dmaIdentifier_e identifier)
+{
+    return dmaDescriptors[DMA_IDENTIFIER_TO_INDEX(identifier)].resourceIndex;
+}
+
+dmaIdentifier_e dmaGetIdentifier(const DMA_Stream_TypeDef* stream)
+{
+    for (int i = 0; i < DMA_LAST_HANDLER; i++) {
+        if (dmaDescriptors[i].ref == stream) {
+            return i + 1;
```

Comment:
isn't there macro to invert DMA_IDENTIFIER_TO_INDEX ?

Reply:
Surprisingly, no...